File: app/src/main/java/pl/karol202/paintplus/tool/shape/polygon/ShapePolygon.java
package pl.karol202.paintplus.tool.shape.polygon;

import android.graphics.*;
import pl.karol202.paintplus.R;
import pl.karol202.paintplus.image.Image;
import pl.karol202.paintplus.tool.shape.OnShapeEditListener;
import pl.karol202.paintplus.tool.shape.Shape;
import pl.karol202.paintplus.tool.shape.ShapeProperties;
import pl.karol202.paintplus.util.Utils;

public class ShapePolygon extends Shape
{
	private int sides;
	private boolean fill;
	private int lineWidth;
	
	private boolean polygonCreated;
	private Point center;
	private float radiusOCC; //Radius of circumscribed circle
	private float angle;
	private Path path;
	
	private Point draggingStart;
	private int draggedIndex;
	private Point centerAtBeginning;
	private float radiusOCCAtBeginning;
	
	public ShapePolygon(Image image, OnShapeEditListener shapeEditListener)
	{
		super(image, shapeEditListener);
		this.sides = 4;
		this.fill = false;
		this.lineWidth = 30;
		
		update();
	}
	
	@Override
	public int getName()
	{
		return R.string.shape_polygon;
	}
	
	@Override
	public int getIcon()
	{
		return R.drawable.ic_shape_polygon_black_24dp;
	}
	
	@Override
	public Class<? extends ShapeProperties> getPropertiesClass()
	{
		return PolygonProperties.class;
	}
	
	public void onTouchStart(int x, int y)
	{
		Point touchPoint = new Point(x, y);
		if(!isInEditMode()) enableEditMode();
		if(!polygonCreated) setCenterPoint(touchPoint);
		else
		{
			float side = (float) (2 * radiusOCC * Math.sin(Math.PI / sides));
			float radiusOIC = (float) (side / (2 * Math.tan(Math.PI / sides)));
			
			float centralAngle = 360f / sides;
			float halfOfCentral = centralAngle / 2;
			float angle = (float) getAngle(touchPoint) - this.angle;
			if(angle < 0) angle += 360;
			float angleMod = angle % centralAngle;
			float a = Math.abs(angleMod - halfOfCentral);
			float centerToSide = Utils.map(a, 0, halfOfCentral, radiusOIC, radiusOCC);
			
			float distanceToCenter = calcDistance(center, x, y);
			float distanceToSide = Math.abs(distanceToCenter - centerToSide);
			
			draggingStart = touchPoint;
			centerAtBeginning = center;
			radiusOCCAtBeginning = radiusOCC;
			if(Math.min(distanceToCenter, distanceToSide) > getMaxTouchDistance()) draggedIndex = -1;
			else if(distanceToCenter < distanceToSide) draggedIndex = 0;
			else draggedIndex = 1;
		}
	}
	
	public void onTouchMove(int x, int y)
	{
		Point current = new Point(x, y);
		if(!polygonCreated) dragRadius(current);
		else drag(current);
	}
	
	public void onTouchStop(int x, int y)
	{
		onTouchMove(x, y);
		polygonCreated = true;
	}
	
	private void drag(Point current)
	{
		if(draggedIndex == 0) dragCenter(current);
		else if(draggedIndex == 1) dragRadius(current);
	}
	
	private void dragCenter(Point current)
	{
		Point delta = new Point(current);
		delta.x -= draggingStart.x;
		delta.y -= draggingStart.y;
		
		Point newCenter = new Point(centerAtBeginning);
		newCenter.offset(delta.x, delta.y);
		setCenterPoint(newCenter);
		
		createPath();
	}
	
	private void dragRadius(Point current)
	{
		if(draggingStart != null)
		{
			double theta = Math.toRadians(Utils.getAngle(center, current));
			float rCurrent = calcDistance(center, current.x, current.y);
			float rDraggingStart = calcDistance(center, draggingStart.x, draggingStart.y);
			float rBeginning = radiusOCCAtBeginning;
			
			float rDelta = rCurrent - rDraggingStart;
			float rResult = rBeginning + rDelta;
			
			float x = (float) (rResult * Math.cos(theta)) + center.x;
			float y = (float) (rResult * Math.sin(theta)) + center.y;
			PointF result = new PointF(x, y);
			getHelpersManager().snapPoint(result);
			
			radiusOCC = calcDistance(center, (int) result.x, (int) result.y);
			
			angle = (float) getAngle(new Point((int) result.x, (int) result.y));
		}
		else
		{
			PointF snapped = new PointF(current);
			getHelpersManager().snapPoint(snapped);
			radiusOCC = calcDistance(center, (int) snapped.x, (int) snapped.y);
			
			angle = (float) getAngle(new Point((int) snapped.x, (int) snapped.y));
		}
		
		createPath();
	}
	
	private double getAngle(Point point)
	{
		double deltaX = point.x - center.x;
		double deltaY = center.y - point.y;
		double ratio = deltaX / deltaY;
		double angleRad = Math.atan(ratio);
		double angleDeg = Math.toDegrees(angleRad);
		if(deltaY < 0) angleDeg += 180;
		if(angleDeg < 0) angleDeg += 360;
		return angleDeg;
	}
	
	private void setCenterPoint(Point point)
	{
		PointF snapped = new PointF(point);
		getHelpersManager().snapPoint(snapped);
		center = new Point((int) snapped.x, (int) snapped.y);
	}
	
	@Override
	public void expandDirtyRect(Rect dirtyRect)
	{
		dirtyRect.left = Math.min(dirtyRect.left, (int) (center.x - radiusOCC - lineWidth));
		dirtyRect.top = Math.min(dirtyRect.top, (int) (center.y - radiusOCC - lineWidth));
		dirtyRect.right = Math.max(dirtyRect.right, (int) (center.x + radiusOCC + lineWidth));
		dirtyRect.bottom = Math.max(dirtyRect.bottom, (int) (center.y + radiusOCC + lineWidth));
	}
	
	@Override
	public void onScreenDraw(Canvas canvas, boolean translucent)
	{
		if(path == null) return;
		updateColor(translucent);
		canvas.drawPath(path, getPaint());
	}
	
	@Override
	public Rect getBoundsOfShape()
	{
		if(path == null) return null;
		RectF rectF = new RectF();
		path.computeBounds(rectF, false);
		Rect rect = new Rect();
		rectF.roundOut(rect);
		rect.left -= lineWidth;
		rect.top -= lineWidth;
		rect.right += lineWidth;
		rect.bottom += lineWidth;
		return rect;
	}
	
	@Override
	public void apply(Canvas imageCanvas)
	{
		if(path == null) return;
		update();
		imageCanvas.drawPath(path, getPaint());
		cleanUp();
	}
	
	@Override
	public void cancel()
	{
		cleanUp();
	}
	
	@Override
	public void update()
	{
		getPaint().setStyle(fill ? Paint.Style.FILL_AND_STROKE : Paint.Style.STROKE);
		getPaint().setStrokeWidth(lineWidth);
		createPath();
		super.update();
	}
	
	@Override
	public void cleanUp()
	{
		polygonCreated = false;
		center = null;
		radiusOCC = -1;
		angle = 0;
		path = null;
		super.cleanUp();
	}
	
	@Override
	public void enableEditMode()
	{
		polygonCreated = false;
		center = null;
		radiusOCC = -1;
		angle = 0;
		path = null;
		super.enableEditMode();
	}
	
	private void createPath()
	{
		if(center == null || radiusOCC == -1) return;
		float central = 360f / sides;
		
		path = new Path();
		for(int i = 0; i < sides; i++)
		{
			float angleRad = (float) Math.toRadians((central * i) + angle);
			float x = center.x + (float) (Math.sin(angleRad) * radiusOCC);
			float y = center.y - (float) (Math.cos(angleRad) * radiusOCC);
			if(i == 0) path.moveTo(x, y);
			else path.lineTo(x, y);
		}
		path.close();
	}
	
	int getSides()
	{
		return sides;
	}
	
	void setSides(int sides)
	{
		if(sides < 3) throw new IllegalArgumentException("Number of sides of polygon cannot be lower than 3.");
		this.sides = sides;
		update();
	}
	
	boolean isFill()
	{
		return fill;
	}
	
	void setFill(boolean fill)
	{
		this.fill = fill;
		update();
	}
	
	int getLineWidth()
	{
		return lineWidth;
	}
	
	void setLineWidth(int lineWidth)
	{
		this.lineWidth = lineWidth;
		update();
	}
}

File: app/src/main/java/pl/karol202/paintplus/tool/shape/polygon/ShapePolygon.kt
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package pl.karol202.paintplus.tool.shape.polygon

import android.content.Context
import android.graphics.*
import androidx.core.graphics.minus
import androidx.core.graphics.plus
import androidx.core.graphics.toPoint
import androidx.core.graphics.toPointF
import pl.karol202.paintplus.R
import pl.karol202.paintplus.helpers.HelpersService
import pl.karol202.paintplus.image.ColorsService
import pl.karol202.paintplus.image.ViewService
import pl.karol202.paintplus.tool.shape.AbstractShape
import pl.karol202.paintplus.tool.shape.Join
import pl.karol202.paintplus.util.*
import pl.karol202.paintplus.util.MathUtils.distance
import pl.karol202.paintplus.util.MathUtils.getAngle
import kotlin.math.*

class ShapePolygon(context: Context,
                   viewService: ViewService,
                   colorsService: ColorsService,
                   private val helpersService: HelpersService) : AbstractShape(context, viewService, colorsService)
{
	private enum class DragType
	{
		CENTER, RADIUS
	}

	private sealed class State
	{
		object Idle : State()

		sealed class Edit : State()
		{
			data class Initial(override val center: Point,
			                   override val radiusOCC: Float,
			                   override val angle: Float) : Edit()

			data class CenterDrag(override val center: Point,
			                      override val radiusOCC: Float,
			                      override val angle: Float,
			                      val dragStart: Point,
			                      val centerAtStart: Point) : Edit()

			data class RadiusDrag(override val center: Point,
			                      override val radiusOCC: Float,
			                      override val angle: Float,
			                      val dragStart: Point,
			                      val radiusAtStart: Float) : Edit()

			abstract val center: Point
			abstract val radiusOCC: Float // Radius of circumscribed circle
			abstract val angle: Float
		}
	}

	override val name get() = R.string.shape_polygon
	override val icon get() = R.drawable.ic_shape_polygon_black_24dp
	override val propertiesClass get() = PolygonProperties::class.java

	var sides by notifying(4).assert { it >= 3 }
	var isFilled by notifying(false)
	var lineWidth by notifying(30)
	var join by notifying(Join.MITTER)

	override val standardPaint by cache({super.standardPaint}, {isFilled}, {lineWidth}, {join}) {
		standardPaint, fill, width, join ->
		Paint(standardPaint).apply {
			style = if(fill) Paint.Style.FILL_AND_STROKE else Paint.Style.STROKE
			strokeWidth = width.toFloat()
			strokeJoin = join.paintJoin
		}
	}
	override val translucentPaint by cache({super.translucentPaint}, {isFilled}, {lineWidth}, {join}) {
		translucentPaint, fill, width, join ->
		Paint(translucentPaint).apply {
			style = if(fill) Paint.Style.FILL_AND_STROKE else Paint.Style.STROKE
			strokeWidth = width.toFloat()
			strokeJoin = join.paintJoin
		}
	}
	private val path by cache({state}) { state ->
		when(state)
		{
			is State.Idle -> null
			is State.Edit -> Path().apply {
				val centralAngle = 360f / sides
				for(i in 0 until sides)
				{
					val angleRad = Math.toRadians((centralAngle * i + state.angle).toDouble()).toFloat()
					val x = state.center.x + (sin(angleRad.toDouble()) * state.radiusOCC).toFloat()
					val y = state.center.y - (cos(angleRad.toDouble()) * state.radiusOCC).toFloat()
					if(i == 0) moveTo(x, y)
					else lineTo(x, y)
				}
				close()
			}
		}
	}

	private var state: State by notifying(State.Idle)

	override val isInEditMode get() = state is State.Edit
	override val bounds get() = path?.computeBounds()?.roundedOut()?.inflated(lineWidth)

	override fun onTouchStart(point: Point)
	{
		state = getInitialState(point)
	}

	private fun getInitialState(point: Point) = when(val state = state)
	{
		is State.Idle -> State.Edit.Initial(snapPoint(point), 0f, 0f)
		is State.Edit ->
		{
			when(getDragType(point, state))
			{
				null -> State.Edit.Initial(state.center, state.radiusOCC, state.angle)
				DragType.CENTER -> State.Edit.CenterDrag(state.center, state.radiusOCC, state.angle, point, state.center)
				DragType.RADIUS -> State.Edit.RadiusDrag(state.center, state.radiusOCC, state.angle, point, state.radiusOCC)
			}
		}
	}

	private fun getDragType(point: Point, state: State.Edit): DragType?
	{
		val side = (2 * state.radiusOCC * sin(Math.PI / sides)).toFloat()
		val radiusOIC = (side / (2 * tan(Math.PI / sides))).toFloat()
		val centralAngle = 360f / sides
		val halfOfCentral = centralAngle / 2
		var angle = getAngle(state.center, point).toFloat() - state.angle
		if(angle < 0) angle += 360f
		val angleMod = angle % centralAngle
		val a = abs(angleMod - halfOfCentral)
		val centerToSide = MathUtils.map(a, 0f, halfOfCentral, radiusOIC, state.radiusOCC)

		val distanceToCenter = distance(state.center, point)
		val distanceToSide = abs(distanceToCenter - centerToSide)
		return when
		{
			min(distanceToCenter, distanceToSide) > maxTouchDistancePx -> null
			distanceToCenter < distanceToSide -> DragType.CENTER
			else -> DragType.RADIUS
		}
	}

	override fun onTouchMove(point: Point)
	{
		state = getEditedState(point) ?: return
	}

	override fun onTouchStop(point: Point)
	{
		state = getEditedState(point) ?: return
	}

	private fun getEditedState(point: Point) = when(val state = state)
	{
		is State.Idle -> null
		is State.Edit.Initial ->
		{
			val snapped = snapPoint(point)
			state.copy(radiusOCC = distance(state.center, snapped),
			           angle = getAngle(state.center, snapped).toFloat() - 90)
		}
		is State.Edit.CenterDrag ->
		{
			val newPoint = state.centerAtStart + (point - state.dragStart)
			state.copy(center = snapPoint(newPoint))
		}
		is State.Edit.RadiusDrag ->
		{
			val theta = Math.toRadians(getAngle(state.center, point))
			val rCurrent = distance(state.center, point)
			val rDraggingStart = distance(state.center, state.dragStart)
			val rBeginning = state.radiusAtStart
			val rDelta = rCurrent - rDraggingStart
			val rResult = rBeginning + rDelta
			val x = (rResult * cos(theta)).toFloat() + state.center.x
			val y = (rResult * sin(theta)).toFloat() + state.center.y
			val result = snapPoint(PointF(x, y))

			state.copy(radiusOCC = distance(state.center.toPointF(), snapPoint(result)),
			           angle = getAngle(state.center.toPointF(), result).toFloat() - 90)
		}
	}

	private fun snapPoint(point: Point) = helpersService.snapPoint(point.toPointF()).toPoint()

	private fun snapPoint(point: PointF) = helpersService.snapPoint(point)

	override fun drawOnLayer(canvas: Canvas, translucent: Boolean)
	{
		val paint = if(translucent) translucentPaint else standardPaint
		canvas.drawPath(path ?: return, paint)
	}

	override fun apply(imageCanvas: Canvas)
	{
		imageCanvas.drawPath(path ?: return, standardPaint)
		cancel()
	}

	override fun cancel()
	{
		state = State.Idle
	}
}
